### PR TITLE
flet: fix out of the box experience

### PR DESCRIPTION
**Pull request by lucasew**
## Description of changes
Right now flet just tries to download the runtime bundle and fails, the objective of this PR is to fix that.

Basically I created this because flet changed a lot since I last tried in #251877

The plan here is to use their monorepo instead of the pypi packages, use a update script on flet-flutter-client and it should propagate to the python libraries so the core and the utilites (from the same repo) would always be synced.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 279936` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>python311Packages.flet</li>
    <li>python311Packages.flet-runtime</li>
    <li>python311Packages.flet-runtime.dist</li>
    <li>python311Packages.flet.dist</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>flet-client-flutter</li>
    <li>flet-client-flutter.pubcache</li>
    <li>python311Packages.flet-core</li>
    <li>python311Packages.flet-core.dist</li>
    <li>python312Packages.flet-core</li>
    <li>python312Packages.flet-core.dist</li>
  </ul>
</details>

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 279936` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>python311Packages.flet</li>
    <li>python311Packages.flet.dist</li>
  </ul>
</details>
<details>
  <summary>8 packages built:</summary>
  <ul>
    <li>flet-client-flutter</li>
    <li>flet-client-flutter.pubcache</li>
    <li>python311Packages.flet-core</li>
    <li>python311Packages.flet-core.dist</li>
    <li>python311Packages.flet-runtime</li>
    <li>python311Packages.flet-runtime.dist</li>
    <li>python312Packages.flet-core</li>
    <li>python312Packages.flet-core.dist</li>
  </ul>
</details>

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 279936` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>flet-client-flutter</li>
    <li>flet-client-flutter.pubcache</li>
    <li>python311Packages.flet</li>
    <li>python311Packages.flet-core</li>
    <li>python311Packages.flet-core.dist</li>
    <li>python311Packages.flet-runtime</li>
    <li>python311Packages.flet-runtime.dist</li>
    <li>python311Packages.flet.dist</li>
    <li>python312Packages.flet-core</li>
    <li>python312Packages.flet-core.dist</li>
  </ul>
</details>

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 279936` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>flet-client-flutter</li>
    <li>flet-client-flutter.pubcache</li>
    <li>python311Packages.flet</li>
    <li>python311Packages.flet-core</li>
    <li>python311Packages.flet-core.dist</li>
    <li>python311Packages.flet-runtime</li>
    <li>python311Packages.flet-runtime.dist</li>
    <li>python311Packages.flet.dist</li>
    <li>python312Packages.flet-core</li>
    <li>python312Packages.flet-core.dist</li>
  </ul>
</details>

**Comment by lucasew** [comment]
How can I fix this ofborg issue?

**Comment by dotlambda** [comment]
> How can I fix this ofborg issue?

I wasn't able to spot the issue. Maybe people experienced with Flutter like @hacker1024 can help.

File: pkgs/by-name/fl/flet-client-flutter/package.nix [hunk 1]
```diff
@@ -19,7 +19,8 @@ flutter.buildFlutterApplication rec {
     rev = "v${version}";
     hash = "sha256-58SMashinWhWZQuo8LTh83eS6ES70jomlGmZtAP9fz4=";
   };
-  sourceRoot = "source/client";
+
+  sourceRoot = "${src.name}/client";
```
Comment (dotlambda) [inline pkgs/by-name/fl/flet-client-flutter/package.nix:23]: This change shouldn't be part of the commit `python311Packages.flet: use shared src from flet-client-flutter as it's a monorepo`.

File: pkgs/development/python-modules/flet-core/default.nix [hunk 2]
```diff
@@ -12,14 +12,10 @@
 
 buildPythonPackage rec {
   pname = "flet-core";
-  version = "0.18.0";
+  inherit (flet-client-flutter) version src;
   format = "pyproject";
```
Comment (dotlambda) [inline pkgs/development/python-modules/flet-core/default.nix:16]:
```suggestion
  pyproject = true;
```

File: pkgs/development/python-modules/flet-runtime/default.nix [hunk 3]
```diff
@@ -10,14 +10,20 @@
...
+
+  postPatch = ''
+    cat <<EOF > src/flet_runtime/_setup_runtime.py
+    import os
+    if 'FLET_VIEW_PATH' not in os.environ:
+      os.environ["FLET_VIEW_PATH"] = "${flet-client-flutter}/bin"
+    EOF
+
+    echo -e "import flet_runtime._setup_runtime\n$(cat src/flet_runtime/__init__.py)" > src/flet_runtime/__init__.py
+  '';
```
Comment (dotlambda) [inline pkgs/development/python-modules/flet-runtime/default.nix:26]: I think an actual patch might make this easier to read, i.e. using `patches`.

**Comment by dotlambda** [comment]
The third commit message should be `python311Packages.flet: add lucasew as maintainer`.

File: pkgs/development/python-modules/flet-runtime/default.nix (hunk 3 above)
Comment (lucasew) [reply pkgs/development/python-modules/flet-runtime/default.nix]: I did an auxiliar file + `substitute`

No inline code, but I had to make sure that the import is the first line

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 279936` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>11 packages built:</summary>
  <ul>
    <li>flet-client-flutter</li>
    <li>flet-client-flutter.debug</li>
    <li>flet-client-flutter.pubcache</li>
    <li>python311Packages.flet</li>
    <li>python311Packages.flet-core</li>
    <li>python311Packages.flet-core.dist</li>
    <li>python311Packages.flet-runtime</li>
    <li>python311Packages.flet-runtime.dist</li>
    <li>python311Packages.flet.dist</li>
    <li>python312Packages.flet-core</li>
    <li>python312Packages.flet-core.dist</li>
  </ul>
</details>

**Comment by lucasew** [comment]
How can I fix ofborg?

What is going on?

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 279936` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>11 packages built:</summary>
  <ul>
    <li>flet-client-flutter</li>
    <li>flet-client-flutter.debug</li>
    <li>flet-client-flutter.pubcache</li>
    <li>python311Packages.flet</li>
    <li>python311Packages.flet-core</li>
    <li>python311Packages.flet-core.dist</li>
    <li>python311Packages.flet-runtime</li>
    <li>python311Packages.flet-runtime.dist</li>
    <li>python311Packages.flet.dist</li>
    <li>python312Packages.flet-core</li>
    <li>python312Packages.flet-core.dist</li>
  </ul>
</details>

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 279936` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>python311Packages.flet</li>
    <li>python311Packages.flet.dist</li>
    <li>python312Packages.flet</li>
    <li>python312Packages.flet.dist</li>
  </ul>
</details>
<details>
  <summary>11 packages built:</summary>
  <ul>
    <li>flet-client-flutter</li>
    <li>flet-client-flutter.debug</li>
    <li>flet-client-flutter.pubcache</li>
    <li>python311Packages.flet-core</li>
    <li>python311Packages.flet-core.dist</li>
    <li>python311Packages.flet-runtime</li>
    <li>python311Packages.flet-runtime.dist</li>
    <li>python312Packages.flet-core</li>
    <li>python312Packages.flet-core.dist</li>
    <li>python312Packages.flet-runtime</li>
    <li>python312Packages.flet-runtime.dist</li>
  </ul>
</details>

File: pkgs/development/compilers/dart/package-source-builders/media_kit_libs_linux/default.nix [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  inherit version src;
+  inherit (src) passthru;
+  doBuild = false;
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p "$out"
+    cp -r '${src}'/* "$out"
+    chmod -R 777 $out/linux/CMakeLists.txt
+    cp ${./CMakeLists.txt} $out/linux/CMakeLists.txt
```
Comment (wegank) [inline pkgs/development/compilers/dart/package-source-builders/media_kit_libs_linux/default.nix:17]: Could this be made into a patch with [bracket comments](https://cmake.org/cmake/help/latest/manual/cmake-language.7.html#bracket-comment)? Or is the diff too big?

Comment (lucasew) [reply pkgs/development/compilers/dart/package-source-builders/media_kit_libs_linux/default.nix:17]: Not sure. The src of that is not pivoted to the root of [1] so it would be harder to maintain a patch this way.

[1] https://github.com/media-kit/media-kit

Comment (lucasew) [reply pkgs/development/compilers/dart/package-source-builders/media_kit_libs_linux/default.nix:17]: BTW this download part is conveniently enclosed with `# -----` things

Comment (lucasew) [reply pkgs/development/compilers/dart/package-source-builders/media_kit_libs_linux/default.nix:17]: This way is better to maintain lol.

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 279936` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>15 packages built:</summary>
  <ul>
    <li>flet-client-flutter</li>
    <li>flet-client-flutter.debug</li>
    <li>flet-client-flutter.pubcache</li>
    <li>python311Packages.flet</li>
    <li>python311Packages.flet-core</li>
    <li>python311Packages.flet-core.dist</li>
    <li>python311Packages.flet-runtime</li>
    <li>python311Packages.flet-runtime.dist</li>
    <li>python311Packages.flet.dist</li>
    <li>python312Packages.flet</li>
    <li>python312Packages.flet-core</li>
    <li>python312Packages.flet-core.dist</li>
    <li>python312Packages.flet-runtime</li>
    <li>python312Packages.flet-runtime.dist</li>
    <li>python312Packages.flet.dist</li>
  </ul>
</details>

File: pkgs/development/compilers/dart/package-source-builders/media_kit_libs_linux/default.nix (hunk 4 above)
Comment (lucasew) [reply pkgs/development/compilers/dart/package-source-builders/media_kit_libs_linux/default.nix]: I did with awk. Much simpler and awk is already in the stdenv so no extra deps.

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 279936` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>15 packages built:</summary>
  <ul>
    <li>flet-client-flutter</li>
    <li>flet-client-flutter.debug</li>
    <li>flet-client-flutter.pubcache</li>
    <li>python311Packages.flet</li>
    <li>python311Packages.flet-core</li>
    <li>python311Packages.flet-core.dist</li>
    <li>python311Packages.flet-runtime</li>
    <li>python311Packages.flet-runtime.dist</li>
    <li>python311Packages.flet.dist</li>
    <li>python312Packages.flet</li>
    <li>python312Packages.flet-core</li>
    <li>python312Packages.flet-core.dist</li>
    <li>python312Packages.flet-runtime</li>
    <li>python312Packages.flet-runtime.dist</li>
    <li>python312Packages.flet.dist</li>
  </ul>
</details>

**Review by wegank — APPROVED** [review]

**Comment by wegank** [comment]
@ofborg build python311Packages.flet

**Comment by lucasew** [comment]
@wegank why is your name on the commits? What did you change?

**Comment by wegank** [comment]
I rebased this PR on staging-next because there are conflicts due to the mass Python update. My name isn't on the commits (check `git log` for that).
